overlord/devicestate/firstboot.go

```diff
@@ -428,19 +428,15 @@ func (m *DeviceManager) importAssertionsFromSeed(isCoreBoot bool) (seed.Seed, er
 	if err != nil {
 		return nil, err
 	}
+
 	modelAssertion := deviceSeed.Model()
 
-	classicModel := modelAssertion.Classic()
-	// FIXME this will not be correct on classic with modes system when
-	// mode is not "run".
-	if release.OnClassic != classicModel {
-		var msg string
-		if classicModel {
-			msg = "cannot seed an all-snaps system with a classic model"
-		} else {
-			msg = "cannot seed a classic system with an all-snaps model"
-		}
-		return nil, errors.New(msg)
+	if release.OnClassic && !modelAssertion.Classic() {
+		return nil, errors.New("cannot seed a classic system with an all-snaps model")
+	}
+
+	if !release.OnClassic && modelAssertion.Classic() && mode != "recover" {
+		return nil, errors.New("can only seed an all-snaps system with a classic model in recovery mode")
```

Comment:
we didn't discuss this but won't this confuse various bits of snapd? do we need to do something special in DeviceContext implementations? also shouldn't we check that the model is actually a hybrid model (there's a kernel in it?)

Reply:
Can you point to something you think might break? I'm not sure I totally follow.

Reply:
These places might be affected for example,

find . -name "*.go"|xargs grep Ctx.Classic|grep -v _test

Reply:
Hrm, I think I understand better after looking around. So, the worry is that we have what looks very much like a core system when we're booted into recovery mode, but the model and device context will report it as a classic system?

If that is the concern, I think we might be fine? I only see a couple places where we check that we're on a classic system (more often we're checking that we're on a core system). I do not think those look problematic.

@alfonsosanchezbeato, you might be able to give some good input here.

Reply:
I think that in the past we have used a combination of `Ctx.IsCoreBoot()`  (UC+hybrid) and `Ctx.Classic()` (classic+hybrid classic) to find out the right things to do. Usage of `Ctx.Classic()`  is happening in `boot`, `configcore` and places in devicestate/snapstate. I think that most of these checks would not happen in this situation (some of them are about updates or installations for instance), but if at one point we start to do factory resets for hybrid, then we might want to start to take into account running model vs target model.

Reply:
I think would be good to leave a TODO, maybe in devicestate/devicecty.go(?) and something in the backlog to revisit this

Reply:
Pushed a change in 627f62b.